Import PyScript Version into Python Runtime

JeffersGlass opened the pull request:
After Ted's PR earlier this week (#952), here's a small-ish PR to access that version in Python as well. It adds `PyScript.__version__` and `PyScript.version_info`:

```python
# namedtuple, mimics sys.version_info
>>> PyScript.version_info
version_info(year=2023, month=2, patch=1, releaselevel='dev')

# human-readable version, similar to what ipython and others do
>>> PyScript.__version__
'2023.02.1.dev'
```

This fixes #546 , and supercedes #833.

antocuni commented on pyscriptjs/src/python/pyscript.py:
```diff
@@ -136,6 +137,39 @@ def write(element_id, value, append=False, exec_id=0):
+            version_dict["releaselevel"] = version_parts[3]
+        except IndexError:
+            version_dict["releaselevel"] = ""
+
```
ah, this seems to contradict my previous comment. But then I don't understand how "normal" versions are supposed to like, with or without the `.final`?

madhur-tandon replied:
I am inclined towards `final` or `release` instead of `""` but have no strong opinions to be honest.

JeffersGlass commented:
Sadly, this breaks in `pyodide.test.ts` in a very strange way that I'm hoping someone can help me figure out.

These lines build fine - if you open a REPL you can see the two new values:

https://github.com/pyscript/pyscript/blob/2740e9817020c66ac9fe34beeeb794e0c6074225/pyscriptjs/src/pyodide.ts#L74-L75

![image](https://user-images.githubusercontent.com/1931111/202479012-44f269bc-92e3-4107-b6d0-5ab81bc81d7e.png)

But `pyodide.test.ts` fails:

```
● PyodideRuntime › should check if runtime is an instance of abstract Runtime

  PythonError: Traceback (most recent call last):
    File "/lib/python3.10/site-packages/_pyodide/_base.py", line 435, in eval_code
      .run(globals, locals)
    File "/lib/python3.10/site-packages/_pyodide/_base.py", line 304, in run
      coroutine = eval(self.code, globals, locals)
    File "<exec>", line 1, in <module>
  NameError: name 'PyScript' is not defined
```

I can't for the life of me figure out why the second `this.run` seems to not be able to access the objects in the initial one, but only in jest.

Any ideas?

JeffersGlass commented on docs/reference/API/__version__.md:
````diff
@@ -0,0 +1,8 @@
+
+```shell
+>>> pyscript.__version__
+'2023.02.1.final'
````
Hmm yeah, 'final' is a bad example here. I'll change it to `2023.02.1.dev` or similar.

FabioRosado commented:
> Sadly, this breaks in `pyodide.test.ts` in a very strange way that I'm hoping someone can help me figure out.
> 
> These lines build fine - if you open a REPL you can see the two new values:
> 
> https://github.com/pyscript/pyscript/blob/2740e9817020c66ac9fe34beeeb794e0c6074225/pyscriptjs/src/pyodide.ts#L74-L75
> 
> ![image](https://user-images.githubusercontent.com/1931111/202479012-44f269bc-92e3-4107-b6d0-5ab81bc81d7e.png)
> 
> But `pyodide.test.ts` fails:
> 
> ```
> ● PyodideRuntime › should check if runtime is an instance of abstract Runtime
> 
>   PythonError: Traceback (most recent call last):
>     File "/lib/python3.10/site-packages/_pyodide/_base.py", line 435, in eval_code
>       .run(globals, locals)
>     File "/lib/python3.10/site-packages/_pyodide/_base.py", line 304, in run
>       coroutine = eval(self.code, globals, locals)
>     File "<exec>", line 1, in <module>
>   NameError: name 'PyScript' is not defined
> ```
> 
> I can't for the life of me figure out why the second `this.run` seems to not be able to access the objects in the initial one, but only in jest.
> 
> Any ideas?

Apologies for the delay Jeff! The pyodide unit tests are importing pyodide and loading it dynamically, I wonder if maybe that's why it's it's complainig that PyScript is not available?

@madhur-tandon perhaps you have more context into this? 🤔

JeffersGlass commented:
Thanks @FabioRosado !

For an extra bit of weirdness, if I change that bit of `loadInterpreter` to:
```ts
this.run(pyscript as string);
logger.info("Dir Results: " + this.run('dir()'))
```
Then this shows up in the jest logs:
```console
console.info
    [pyscript/pyodide] Dir Results: ['__annotations__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', 'pyversion', 'version_info']
```
So `dir()` is running fine... but with none of the symbols/globals that should be there after running pyscript.py? Very odd.

JeffersGlass commented:
Ok some further oddities which make me think it's something odd about async interactions between pyodide and jest. Both of the following examples work when building for the browser, but...

This (the above example) breaks jest:
```ts
//main.ts
async afterRuntimeLoad(){
    ...
    await runtime.loadInterpreter();
    ...
    }

//pyodide.ts
async loadInterpreter(){
    ...
    this.run(pyscript as string);
    this.run(`PyScript.set_version_info('${version}')`)
```

BUT this works in jest:
```ts
//main.ts
async afterRuntimeLoad(){
    ...
    await runtime.loadInterpreter();
    runtime.run(`PyScript.set_version_info('${version}')`)
    }

//pyodide.ts
async loadInterpreter(){
    ...
    this.run(pyscript as string);
    ....
    }
```
🤯

FabioRosado commented:
That's interesting, the pyodide tests are doing this:

https://github.com/pyscript/pyscript/blob/0e1c396d7c29d06cafb5fa4dca9901c2a2493aaf/pyscriptjs/tests/unit/pyodide.test.ts#L41

I wonder if this is related? 🤔   Wonder if perhaps runtime didn't finish load (not sure if this even makes sense haha)

JeffersGlass commented:
I figured it out!

jest is using a mocked version of `pyscript.py` for the tests:
https://github.com/JeffersGlass/pyscript/blob/803988342304d576425d48fd79ce3ebc50047226/pyscriptjs/jest.config.js#L19-L21

And that mock is.... empty!
https://github.com/JeffersGlass/pyscript/blob/803988342304d576425d48fd79ce3ebc50047226/pyscriptjs/__mocks__/fileMock.js#L12

So of course "PyScript isn't defined". 

I think the solution is: change `fileMock.js` so that it copies and exports the contents of PyScript.py instead of "". I think this gives us the best approximation of the in-browser environment going forward. I'm just not sure the best way to do that... but I'll figure it out. Unless @madhur-tandon has suggestions?

FabioRosado commented:
Oh dang! That's a great find, awesome work!

Copying over seems like a good idea or we could copy only the needed methods if that works, what do you think?

madhur-tandon commented:
@JeffersGlass @FabioRosado 

Sorry for the late reply. Yeah, I did the mocking thing with an empty string because `jest`
threw errors when trying to import a `.py` script.

The project overall builds without failure since our `rollup.config.js` has

https://github.com/pyscript/pyscript/blob/0e1c396d7c29d06cafb5fa4dca9901c2a2493aaf/pyscriptjs/rollup.config.js#L44-L46

aka it is interpreted as a raw string. If we can somehow apply this setting to `jest`, that will be the best solution IMO but a JS expert could probably weigh in on how to do that.

Else, it seems like copying relevant stuff that is needed into the mocked string should be decently sufficient I guess.

FabioRosado commented:
> @JeffersGlass @FabioRosado
> 
> Sorry for the late reply. Yeah, I did the mocking thing with an empty string because `jest` threw errors when trying to import a `.py` script.
> 
> The project overall builds without failure since our `rollup.config.js` has
> 
> https://github.com/pyscript/pyscript/blob/0e1c396d7c29d06cafb5fa4dca9901c2a2493aaf/pyscriptjs/rollup.config.js#L44-L46
> 
> aka the contents of `.py` imports are interpreted as raw strings. If we can somehow apply this setting to `jest`, that will be the best solution IMO but a JS ninja 🥷 could probably weigh in on how to do that.
> 
> Else, it seems like copying relevant stuff that is needed into the mocked string should be decently sufficient I guess.

You can get jest to accept .py files with the config:

```
    moduleFileExtensions: ["js", "mjs", "cjs", "jsx", "ts", "tsx", "json", "node", "py"],
```

But then the tests explode because of the python imports

```
SyntaxError: /Users/fabiorosado/github/pyscript/pyscriptjs/src/python/pyscript.py: Unexpected token, expected "from" (2:0)

      1 | import ast
    > 2 | import asyncio
```

Maybe we go for the easier path and expand the mocked file 😄

JeffersGlass commented:
> Maybe we go for the easier path and expand the mocked file

Thanks y'all - for now, I've added `__mocks__/_pyscript.js` as an explicit mock of `pyscript.py` that jest can use at test time. It loads the contents of "pyscript.py" and exports them as a string.

All other .py files are still caught by the original mapping and mocked with `fileMock.js` (i.e. as an empty string), but I've added a console warning there that files are being mocked as "" for the next time someone runs across this.

madhur-tandon commented on pyscriptjs/__mocks__/fileMock.js:
```diff
@@ -1,12 +1,16 @@
- * since importing of `.py` files isn't usually supported
- * inside JS/TS files.
+ * this file mocks python files that are not explicitly
+ * match by a regex in jest.config.js importing of
```
Very minor nitpicks, and they probably don't matter but would be nice to have.

`explicitly match` --> `explicitly matched`.

`jest.config.js importing of` -->. `jest.config.js. Importing of`

`is import from that` --> `is imported from that`

`module.exports = ""` --> `module.exports = "";` (missing semicolon)

JeffersGlass replied:
Genuinely very much appreciate these. Thank you!

FabioRosado commented on pyscriptjs/__mocks__/_pyscript.js:
```diff
@@ -0,0 +1,14 @@
+ */
+
+const fs = require('fs');
+module.exports = fs.readFileSync('./src/python/pyscript.py', 'utf8');
```
Apologies a bit late to the party but... NICE! 😄
